mpi-proxy-split/mpi-wrappers/mpi_file_wrappers.cpp

```diff
@@ -42,19 +46,295 @@ USER_DEFINED_WRAPPER(int, File_open, (MPI_Comm) comm, (const char *) filename,
   JUMP_TO_LOWER_HALF(lh_info.fsaddr);
   retval = NEXT_FUNC(File_open)(realComm, filename, amode, info, fh);
   RETURN_TO_UPPER_HALF();
+  if (retval == MPI_SUCCESS) {
+    // Add file to virtual to real mapping
+    MPI_File virtFile = ADD_NEW_FILE(*fh);
+    *fh = virtFile;
+
+    // Save parameters
+    OpenFileParameters params;
+    params._mode = amode;
+    params._comm = comm;
+    params._info = info;
+    strcpy(params._filepath, filename);
+
+    // If view isn't set, we can't restore any view
+    params._viewSet = 0;
+
+    // Save this file handle to the global map
+    g_params_map[virtFile] = params;
+  }
+  DMTCP_PLUGIN_ENABLE_CKPT();
+  return retval;
+}
+
+USER_DEFINED_WRAPPER(int, File_get_atomicity, (MPI_File) fh, (int*) flag)
+{
+  int retval;
+  DMTCP_PLUGIN_DISABLE_CKPT();
+  MPI_File realFile = VIRTUAL_TO_REAL_FILE(fh);
+  JUMP_TO_LOWER_HALF(lh_info.fsaddr);
+  retval = NEXT_FUNC(File_get_atomicity)(realFile, flag);
+  RETURN_TO_UPPER_HALF();
+  DMTCP_PLUGIN_ENABLE_CKPT();
+  return retval;
+}
+
+USER_DEFINED_WRAPPER(int, File_set_atomicity, (MPI_File) fh, (int) flag)
+{
+  int retval;
+  DMTCP_PLUGIN_DISABLE_CKPT();
+  MPI_File realFile = VIRTUAL_TO_REAL_FILE(fh);
+  JUMP_TO_LOWER_HALF(lh_info.fsaddr);
+  retval = NEXT_FUNC(File_set_atomicity)(realFile, flag);
+  RETURN_TO_UPPER_HALF();
+  DMTCP_PLUGIN_ENABLE_CKPT();
+  return retval;
+}
+
+USER_DEFINED_WRAPPER(int, File_set_size, (MPI_File) fh, (MPI_Offset) size)
+{
+  int retval;
+  DMTCP_PLUGIN_DISABLE_CKPT();
+  MPI_File realFile = VIRTUAL_TO_REAL_FILE(fh);
+  JUMP_TO_LOWER_HALF(lh_info.fsaddr);
+  retval = NEXT_FUNC(File_set_size)(realFile, size);
+  RETURN_TO_UPPER_HALF();
+  DMTCP_PLUGIN_ENABLE_CKPT();
+  return retval;
+}
+
+USER_DEFINED_WRAPPER(int, File_get_size, (MPI_File) fh, (MPI_Offset *) size)
+{
+  int retval;
+  DMTCP_PLUGIN_DISABLE_CKPT();
+  MPI_File realFile = VIRTUAL_TO_REAL_FILE(fh);
+  JUMP_TO_LOWER_HALF(lh_info.fsaddr);
+  retval = NEXT_FUNC(File_get_size)(realFile, size);
+  RETURN_TO_UPPER_HALF();
+  DMTCP_PLUGIN_ENABLE_CKPT();
+  return retval;
+}
+
+USER_DEFINED_WRAPPER(int, File_set_view, (MPI_File) fh, (MPI_Offset) disp,
+                     (MPI_Datatype) etype, (MPI_Datatype) filetype,
+                     (const char*) datarep, (MPI_Info) info)
+{
+  int retval;
+  DMTCP_PLUGIN_DISABLE_CKPT();
+  MPI_File realFile = VIRTUAL_TO_REAL_FILE(fh);
+  MPI_Datatype realEtype = VIRTUAL_TO_REAL_TYPE(etype);
+  MPI_Datatype realFtype = VIRTUAL_TO_REAL_TYPE(filetype);
+  JUMP_TO_LOWER_HALF(lh_info.fsaddr);
+  retval = NEXT_FUNC(File_set_view)(realFile, disp, realEtype, realFtype,
+                                    datarep, info);
+  RETURN_TO_UPPER_HALF();
+
+  // Normally, we wait until checkpoint time to update any file characteristics
+  // However, we can't call set_view if there is no valid view
```

Comment:
set_view or get_view?

Reply:
set_view
get_view completes, but gives garbage results

Reply:
You'll be calling get_view during a checkpoint, and set_view during the restore, right?

Could you also update the comment clarifying that you will update the file characteristics in the global map, not in the MPI_File in the lower half?